File: tests/human/x/pl/left_join.pl
:- initialization(main).

customers([
  customer{id:1, name:'Alice'},
  customer{id:2, name:'Bob'}
]).

orders([
  order{id:100, customerId:1, total:250},
  order{id:101, customerId:3, total:80}
]).

find_customer(Id, Customers, Cust) :-
    (member(C, Customers), C.id = Id -> Cust = C ; Cust = nil).

left_join([], _, []).
left_join([O|Os], Cs, [entry{orderId:O.id, customer:Cust, total:O.total}|Rs]) :-
    find_customer(O.customerId, Cs, Cust),
    left_join(Os, Cs, Rs).

format_customer(nil, '<nil>').
format_customer(C, Str) :-
    C \= nil,
    format(atom(Str), 'map[id:~w name:~w]', [C.id, C.name]).

main :-
    customers(Cs), orders(Os), left_join(Os, Cs, Result),
    writeln('--- Left Join ---'),
    forall(member(E, Result),
        (format_customer(E.customer, CS),
         format('Order ~w customer ~w total ~w~n', [E.orderId, CS, E.total]))),
    halt.

File: tests/human/x/fs/left_join.fs
open System

type Customer = { id:int; name:string }
type Order = { id:int; customerId:int; total:int }

let customers = [
    { id = 1; name = "Alice" }
    { id = 2; name = "Bob" }
]

let orders = [
    { id = 100; customerId = 1; total = 250 }
    { id = 101; customerId = 3; total = 80 }
]

let result =
    [ for o in orders do
        let customer = customers |> List.tryFind (fun c -> c.id = o.customerId)
        yield {| orderId = o.id; customer = customer; total = o.total |} ]

printfn "--- Left Join ---"
for entry in result do
    printfn "Order %d customer %A total %d" entry.orderId entry.customer entry.total
